File: apps/setup/src/contracts/wizard.ts
/**
 * EP-035 M1 SetupWizard contract (SPEC-004 / SPEC-016).
 *
 * The wizard models STATE, not visual progress. A page being visited or
 * rendered never implies a step is complete, a provider is configured,
 * or a resource is healthy. Two structurally distinct step statuses
 * exist on purpose:
 *
 *   COMPLETE_LOCAL  - local checkpoint only (LOCAL_PROGRESS_SAVED)
 *   VERIFIED        - remote effect verified (REMOTE_EFFECT_VERIFIED)
 *
 * A COMPLETE_LOCAL step is never VERIFIED without an explicit remote
 * verification record, and a cached completed-step flag can never
 * satisfy readiness. Transitions are typed and validated: invalid
 * leaps (NOT_STARTED -> COMPLETED, FAILED -> COMPLETED without
 * recovery) are rejected.
 */

import {
  assertEnum,
  assertInt,
  assertNonNegativeInt,
  assertObject,
  assertString,
  assertUuid,
  rejectUnknownFields,
} from "./validate";
import { ErrorCode, Spec006Error } from "./errors";

export const WIZARD_STATES = [
  "NOT_STARTED",
  "IN_PROGRESS",
  "BLOCKED",
  "FAILED",
  "RECOVERY_REQUIRED",
  "COMPLETED",
] as const;
export type WizardState = (typeof WIZARD_STATES)[number];
const WIZARD_STATE_SET: ReadonlySet<WizardState> = new Set(WIZARD_STATES);

export const WIZARD_STEPS = [
  "DEPLOYMENT_CHOICE",
  "HARDWARE_PROFILE",
  "OWNER_BOOTSTRAP",
  "RECOVERY_MATERIAL",
  "EDGE_ENROLLMENT",
  "DISCOVERY",
  "INTEGRATION_REVIEW",
  "PLAN_REVIEW",
] as const;
export type WizardStep = (typeof WIZARD_STEPS)[number];
const WIZARD_STEP_SET: ReadonlySet<WizardStep> = new Set(WIZARD_STEPS);

export const WIZARD_STEP_STATUSES = [
  "PENDING",
  "IN_PROGRESS",
  "BLOCKED",
  "FAILED",
  "COMPLETE_LOCAL",
  "VERIFIED",
] as const;
export type WizardStepStatus = (typeof WIZARD_STEP_STATUSES)[number];
const WIZARD_STEP_STATUS_SET: ReadonlySet<WizardStepStatus> = new Set(
  WIZARD_STEP_STATUSES,
);

/** Allowed whole-wizard state transitions (typed, validated). */
const WIZARD_TRANSITIONS: Readonly<
  Record<WizardState, ReadonlySet<WizardState>>
> = {
  NOT_STARTED: new Set<WizardState>(["IN_PROGRESS"]),
  IN_PROGRESS: new Set<WizardState>([
    "BLOCKED",
    "FAILED",
    "RECOVERY_REQUIRED",
    "COMPLETED",
  ]),
  BLOCKED: new Set<WizardState>(["IN_PROGRESS", "RECOVERY_REQUIRED"]),
  FAILED: new Set<WizardState>(["RECOVERY_REQUIRED", "IN_PROGRESS"]),
  RECOVERY_REQUIRED: new Set<WizardState>(["IN_PROGRESS"]),
  COMPLETED: new Set<WizardState>([]),
};

/** Allowed per-step status transitions (typed, validated). */
const STEP_STATUS_TRANSITIONS: Readonly<
  Record<WizardStepStatus, ReadonlySet<WizardStepStatus>>
> = {
  PENDING: new Set<WizardStepStatus>(["IN_PROGRESS"]),
  IN_PROGRESS: new Set<WizardStepStatus>([
    "BLOCKED",
    "FAILED",
    "COMPLETE_LOCAL",
  ]),
  BLOCKED: new Set<WizardStepStatus>(["IN_PROGRESS"]),
  FAILED: new Set<WizardStepStatus>(["IN_PROGRESS"]),
  COMPLETE_LOCAL: new Set<WizardStepStatus>(["VERIFIED"]),
  VERIFIED: new Set<WizardStepStatus>([]),
};

export interface RemoteVerificationShape {
  verified_at_unix_s: number;
  verifier: string;
}

export interface WizardStepRecordShape {
  step: WizardStep;
  status: WizardStepStatus;
  last_transition_at_unix_s: number;
  verification?: RemoteVerificationShape | null;
}

export interface SetupWizardStateShape {
  state: WizardState;
  current_step: WizardStep;
  steps: Array<WizardStepRecordShape>;
  correlation_id: string;
  updated_at_unix_s: number;
}

export interface WizardBeginRequestShape {
  correlation_id: string;
}

export interface WizardAdvanceRequestShape {
  correlation_id: string;
  step: WizardStep;
  to_status: WizardStepStatus;
}

export interface WizardVerifyRequestShape {
  correlation_id: string;
  step: WizardStep;
  verification: RemoteVerificationShape;
}

const REMOTE_VERIFICATION_FIELDS = new Set<string>([
  "verified_at_unix_s",
  "verifier",
]);

function parseRemoteVerification(
  value: unknown,
  what: string,
): RemoteVerificationShape {
  const obj = assertObject(value, what);
  rejectUnknownFields(obj, REMOTE_VERIFICATION_FIELDS, what);
  return {
    verified_at_unix_s: assertNonNegativeInt(
      obj["verified_at_unix_s"],
      `${what}.verified_at_unix_s`,
    ),
    verifier: assertNonEmptyVerifier(obj["verifier"], `${what}.verifier`),
  };
}

function assertNonEmptyVerifier(value: unknown, what: string): string {
  const s = assertString(value, what);
  if (s.length === 0) {
    throw new Spec006Error(ErrorCode.Validation, `${what} must not be empty`);
  }
  return s;
}

const STEP_RECORD_FIELDS = new Set<string>([
  "step",
  "status",
  "last_transition_at_unix_s",
  "verification",
]);

export class WizardStepRecord {
  readonly step: WizardStep;
  readonly status: WizardStepStatus;
  readonly last_transition_at_unix_s: number;
  readonly verification: RemoteVerificationShape | undefined;

  /** Constructed only by wizard state operations; exposed for deterministic tests. */
  constructor(
    step: WizardStep,
    status: WizardStepStatus,
    lastTransitionAtUnixS: number,
    verification: RemoteVerificationShape | undefined,
  ) {
    this.step = step;
    this.status = status;
    this.last_transition_at_unix_s = lastTransitionAtUnixS;
    this.verification = verification;
  }

  static parse(value: unknown, what = "wizard step record"): WizardStepRecord {
    const obj = assertObject(value, what);
    rejectUnknownFields(obj, STEP_RECORD_FIELDS, what);
    const step = assertEnum(obj["step"], WIZARD_STEP_SET, `${what}.step`);
    const status = assertEnum(
      obj["status"],
      WIZARD_STEP_STATUS_SET,
      `${what}.status`,
    );
    const lastTransition = assertNonNegativeInt(
      obj["last_transition_at_unix_s"],
      `${what}.last_transition_at_unix_s`,
    );
    const verificationRaw = obj["verification"];
    if (verificationRaw !== undefined && verificationRaw !== null) {
      const verification = parseRemoteVerification(
        verificationRaw,
        `${what}.verification`,
      );
      if (status !== "VERIFIED") {
        throw new Spec006Error(
          ErrorCode.Validation,
          `${what} carries a verification record but status is ${status}; only VERIFIED steps may carry one`,
        );
      }
      return new WizardStepRecord(step, status, lastTransition, verification);
    }
    if (status === "VERIFIED") {
      throw new Spec006Error(
        ErrorCode.Verification,
        `${what} is VERIFIED but has no verification record`,
      );
    }
    return new WizardStepRecord(step, status, lastTransition, undefined);
  }

  toJSON(): WizardStepRecordShape {
    return {
      step: this.step,
      status: this.status,
      last_transition_at_unix_s: this.last_transition_at_unix_s,
      ...(this.verification === undefined
        ? {}
        : { verification: this.verification }),
    };
  }
}

export function isValidWizardStateTransition(
  from: WizardState,
  to: WizardState,
): boolean {
  const allowed = WIZARD_TRANSITIONS[from];
  return allowed !== undefined && allowed.has(to);
}

export function isValidStepStatusTransition(
  from: WizardStepStatus,
  to: WizardStepStatus,
): boolean {
  const allowed = STEP_STATUS_TRANSITIONS[from];
  return allowed !== undefined && allowed.has(to);
}

const WIZARD_STATE_FIELDS = new Set<string>([
  "state",
  "current_step",
  "steps",
  "correlation_id",
  "updated_at_unix_s",
]);

export class SetupWizardState {
  readonly state: WizardState;
  readonly current_step: WizardStep;
  readonly steps: ReadonlyArray<WizardStepRecord>;
  readonly correlation_id: string;
  readonly updated_at_unix_s: number;

  private constructor(
    state: WizardState,
    currentStep: WizardStep,
    steps: ReadonlyArray<WizardStepRecord>,
    correlationId: string,
    updatedAtUnixS: number,
  ) {
    this.state = state;
    this.current_step = currentStep;
    this.steps = steps;
    this.correlation_id = correlationId;
    this.updated_at_unix_s = updatedAtUnixS;
  }

  static parse(value: unknown, what = "setup wizard state"): SetupWizardState {
    const obj = assertObject(value, what);
    rejectUnknownFields(obj, WIZARD_STATE_FIELDS, what);
    const state = assertEnum(obj["state"], WIZARD_STATE_SET, `${what}.state`);
    const currentStep = assertEnum(
      obj["current_step"],
      WIZARD_STEP_SET,
      `${what}.current_step`,
    );
    const stepsRaw = obj["steps"];
    if (!Array.isArray(stepsRaw)) {
      throw new Spec006Error(
        ErrorCode.Validation,
        `${what}.steps must be an array`,
      );
    }
    const steps = stepsRaw.map((entry) =>
      WizardStepRecord.parse(entry, `${what}.steps entry`),
    );
    const seen = new Set<WizardStep>();
    for (const step of steps) {
      if (seen.has(step.step)) {
        throw new Spec006Error(
          ErrorCode.Validation,
          `${what}.steps contains duplicate step '${step.step}'`,
        );
      }
      seen.add(step.step);
    }
    for (const stepId of WIZARD_STEPS) {
      if (!seen.has(stepId)) {
        throw new Spec006Error(
          ErrorCode.Validation,
          `${what}.steps is missing step '${stepId}'`,
        );
      }
    }
    return new SetupWizardState(
      state,
      currentStep,
      steps,
      assertUuid(obj["correlation_id"], `${what}.correlation_id`),
      assertNonNegativeInt(
        obj["updated_at_unix_s"],
        `${what}.updated_at_unix_s`,
      ),
    );
  }

  /** Create the canonical NOT_STARTED wizard with every step PENDING. */
  static notStarted(correlationId: string, atUnixS: number): SetupWizardState {
    const steps = WIZARD_STEPS.map(
      (step) => new WizardStepRecord(step, "PENDING", atUnixS, undefined),
    );
    return new SetupWizardState(
      "NOT_STARTED",
      WIZARD_STEPS[0],
      steps,
      correlationId,
      atUnixS,
    );
  }

  stepRecord(step: WizardStep): WizardStepRecord {
    const record = this.steps.find((entry) => entry.step === step);
    if (record === undefined) {
      throw new Spec006Error(
        ErrorCode.Internal,
        `wizard has no record for step '${step}'`,
      );
    }
    return record;
  }

  /**
   * Advance the whole wizard state. COMPLETED additionally requires
   * every step to be VERIFIED: a wizard with unverified steps can never
   * be marked complete.
   */
  advance(toState: WizardState, atUnixS: number): SetupWizardState {
    if (!isValidWizardStateTransition(this.state, toState)) {
      throw new Spec006Error(
        ErrorCode.Policy,
        `invalid wizard transition ${this.state} -> ${toState}`,
        this.correlation_id,
      );
    }
    if (toState === "COMPLETED") {
      const unverified = this.steps.filter(
        (entry) => entry.status !== "VERIFIED",
      );
      if (unverified.length > 0) {
        throw new Spec006Error(
          ErrorCode.Policy,
          `wizard cannot complete with unverified steps: ${unverified
            .map((entry) => entry.step)
            .join(",")}`,
          this.correlation_id,
        );
      }
    }
    return new SetupWizardState(
      toState,
      this.current_step,
      this.steps,
      this.correlation_id,
      atUnixS,
    );
  }

  /** Advance one step's status (typed transition; VERIFIED needs a record). */
  advanceStep(
    step: WizardStep,
    toStatus: WizardStepStatus,
    atUnixS: number,
    verification?: RemoteVerificationShape,
  ): SetupWizardState {
    const record = this.stepRecord(step);
    if (toStatus !== "VERIFIED" && verification !== undefined) {
      throw new Spec006Error(
        ErrorCode.Validation,
        `step ${step} cannot carry a verification record with status ${toStatus}`,
        this.correlation_id,
      );
    }
    if (!isValidStepStatusTransition(record.status, toStatus)) {
      throw new Spec006Error(
        ErrorCode.Policy,
        `invalid step transition ${step}: ${record.status} -> ${toStatus}`,
        this.correlation_id,
      );
    }
    if (toStatus === "VERIFIED" && verification === undefined) {
      throw new Spec006Error(
        ErrorCode.Verification,
        `step ${step} cannot become VERIFIED without a verification record`,
        this.correlation_id,
      );
    }
    const updated: Array<WizardStepRecord> = this.steps.map((entry) => {
      if (entry.step !== step) {
        return entry;
      }
      return new WizardStepRecord(step, toStatus, atUnixS, verification);
    });
    return new SetupWizardState(
      this.state,
      step,
      updated,
      this.correlation_id,
      atUnixS,
    );
  }

  toJSON(): SetupWizardStateShape {
    return {
      state: this.state,
      current_step: this.current_step,
      steps: this.steps.map((entry) => entry.toJSON()),
      correlation_id: this.correlation_id,
      updated_at_unix_s: this.updated_at_unix_s,
    };
  }
}

const WIZARD_BEGIN_FIELDS = new Set<string>(["correlation_id"]);
const WIZARD_ADVANCE_FIELDS = new Set<string>([
  "correlation_id",
  "step",
  "to_status",
]);
const WIZARD_VERIFY_FIELDS = new Set<string>([
  "correlation_id",
  "step",
  "verification",
]);

export class WizardAdvanceRequest {
  readonly correlation_id: string;
  readonly step: WizardStep;
  readonly to_status: WizardStepStatus;

  private constructor(
    correlationId: string,
    step: WizardStep,
    toStatus: WizardStepStatus,
  ) {
    this.correlation_id = correlationId;
    this.step = step;
    this.to_status = toStatus;
  }

  static parse(value: unknown): WizardAdvanceRequest {
    const obj = assertObject(value, "wizard advance request");
    rejectUnknownFields(obj, WIZARD_ADVANCE_FIELDS, "wizard advance request");
    return new WizardAdvanceRequest(
      assertUuid(
        obj["correlation_id"],
        "wizard advance request.correlation_id",
      ),
      assertEnum(obj["step"], WIZARD_STEP_SET, "wizard advance request.step"),
      assertEnum(
        obj["to_status"],
        WIZARD_STEP_STATUS_SET,
        "wizard advance request.to_status",
      ),
    );
  }
}

export class WizardVerifyRequest {
  readonly correlation_id: string;
  readonly step: WizardStep;
  readonly verification: RemoteVerificationShape;

  private constructor(
    correlationId: string,
    step: WizardStep,
    verification: RemoteVerificationShape,
  ) {
    this.correlation_id = correlationId;
    this.step = step;
    this.verification = verification;
  }

  static parse(value: unknown): WizardVerifyRequest {
    const obj = assertObject(value, "wizard verify request");
    rejectUnknownFields(obj, WIZARD_VERIFY_FIELDS, "wizard verify request");
    return new WizardVerifyRequest(
      assertUuid(obj["correlation_id"], "wizard verify request.correlation_id"),
      assertEnum(obj["step"], WIZARD_STEP_SET, "wizard verify request.step"),
      parseRemoteVerification(
        obj["verification"],
        "wizard verify request.verification",
      ),
    );
  }
}

/**
 * Provider-neutral SetupWizard port. M1 declares the boundary; the real
 * durable implementation is owned by M2+.
 */
export interface SetupWizardPort {
  begin(request: WizardBeginRequestShape): SetupWizardState;
  advance(request: WizardAdvanceRequestShape): SetupWizardState;
  verifyRemote(request: WizardVerifyRequestShape): SetupWizardState;
}

File: crates/nexus-setup/src/wizard.rs
//! EP-035 SetupWizard state machine (SPEC-004 / SPEC-016).
//!
//! The wizard models STATE, not visual progress. A page being visited
//! or rendered never implies a step is complete, a provider is
//! configured, or a resource is healthy. COMPLETE_LOCAL is a local
//! checkpoint (LOCAL_PROGRESS_SAVED); VERIFIED requires an explicit
//! remote verification record (REMOTE_EFFECT_VERIFIED). Transitions are
//! typed and validated: invalid leaps (NOT_STARTED -> COMPLETED,
//! FAILED -> COMPLETED without recovery) are rejected with POLICY.

use std::collections::BTreeMap;

use nexus_domain::CorrelationId;
use serde::{Deserialize, Serialize};

use crate::error::{SetupError, SetupResult};
use crate::vocabulary::{WizardState, WizardStep, WizardStepStatus};

/// Allowed whole-wizard state transitions.
pub fn is_valid_wizard_transition(from: WizardState, to: WizardState) -> bool {
    use WizardState::*;
    matches!(
        (from, to),
        (NotStarted, InProgress)
            | (InProgress, Blocked | Failed | RecoveryRequired | Completed)
            | (Blocked, InProgress | RecoveryRequired)
            | (Failed, RecoveryRequired | InProgress)
            | (RecoveryRequired, InProgress)
    )
}

/// Allowed per-step status transitions.
pub fn is_valid_step_transition(from: WizardStepStatus, to: WizardStepStatus) -> bool {
    use WizardStepStatus::*;
    matches!(
        (from, to),
        (Pending, InProgress)
            | (InProgress, Blocked | Failed | CompleteLocal)
            | (Blocked, InProgress)
            | (Failed, InProgress)
            | (CompleteLocal, Verified)
    )
}

/// Remote verification record. Only a VERIFIED step carries one.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RemoteVerification {
    pub verified_at_unix_s: u64,
    pub verifier: String,
}

/// Per-step wizard record.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WizardStepRecord {
    pub step: WizardStep,
    pub status: WizardStepStatus,
    pub last_transition_at_unix_s: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub verification: Option<RemoteVerification>,
}

impl WizardStepRecord {
    pub fn new(
        step: WizardStep,
        status: WizardStepStatus,
        last_transition_at_unix_s: u64,
        verification: Option<RemoteVerification>,
    ) -> SetupResult<Self> {
        if verification.is_some() && status != WizardStepStatus::Verified {
            return Err(SetupError::validation(format!(
                "step {} carries a verification record but status is {}",
                step, status
            )));
        }
        if status == WizardStepStatus::Verified && verification.is_none() {
            return Err(SetupError::verification(format!(
                "step {} is VERIFIED but has no verification record",
                step
            )));
        }
        Ok(Self {
            step,
            status,
            last_transition_at_unix_s,
            verification,
        })
    }
}

/// All steps owned by the setup wizard, in canonical order.
pub const ALL_WIZARD_STEPS: [WizardStep; 8] = [
    WizardStep::DeploymentChoice,
    WizardStep::HardwareProfile,
    WizardStep::OwnerBootstrap,
    WizardStep::RecoveryMaterial,
    WizardStep::EdgeEnrollment,
    WizardStep::Discovery,
    WizardStep::IntegrationReview,
    WizardStep::PlanReview,
];

/// SetupWizardState: the whole wizard, every step, typed transitions.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SetupWizardState {
    pub state: WizardState,
    pub current_step: WizardStep,
    pub steps: Vec<WizardStepRecord>,
    pub correlation: CorrelationId,
    pub updated_at_unix_s: u64,
}

impl SetupWizardState {
    /// Canonical NOT_STARTED wizard with every step PENDING.
    pub fn not_started(correlation: CorrelationId, at_unix_s: u64) -> Self {
        let steps = ALL_WIZARD_STEPS
            .iter()
            .map(|step| WizardStepRecord {
                step: *step,
                status: WizardStepStatus::Pending,
                last_transition_at_unix_s: at_unix_s,
                verification: None,
            })
            .collect();
        Self {
            state: WizardState::NotStarted,
            current_step: WizardStep::DeploymentChoice,
            steps,
            correlation,
            updated_at_unix_s: at_unix_s,
        }
    }

    /// Validate the invariant: exactly the canonical step set, no
    /// duplicates, VERIFIED records consistent.
    pub fn validate(&self) -> SetupResult<()> {
        let mut seen: BTreeMap<WizardStep, usize> = BTreeMap::new();
        for record in &self.steps {
            *seen.entry(record.step).or_insert(0) += 1;
        }
        for step in ALL_WIZARD_STEPS {
            let count = seen.get(&step).copied().unwrap_or(0);
            if count != 1 {
                return Err(SetupError::validation(format!(
                    "wizard steps must contain exactly one record per step; {} has {}",
                    step, count
                )));
            }
        }
        for record in &self.steps {
            if record.verification.is_some() && record.status != WizardStepStatus::Verified {
                return Err(SetupError::validation(format!(
                    "step {} carries verification with status {}",
                    record.step, record.status
                )));
            }
        }
        Ok(())
    }

    pub fn step_record(&self, step: WizardStep) -> Option<&WizardStepRecord> {
        self.steps.iter().find(|record| record.step == step)
    }

    /// Advance the whole wizard. COMPLETED requires every step VERIFIED.
    pub fn advance(&self, to_state: WizardState, at_unix_s: u64) -> SetupResult<Self> {
        if !is_valid_wizard_transition(self.state, to_state) {
            return Err(SetupError::policy(format!(
                "invalid wizard transition {} -> {}",
                self.state, to_state
            )));
        }
        if to_state == WizardState::Completed {
            let unverified: Vec<WizardStep> = self
                .steps
                .iter()
                .filter(|record| record.status != WizardStepStatus::Verified)
                .map(|record| record.step)
                .collect();
            if !unverified.is_empty() {
                let names: Vec<String> = unverified.iter().map(|s| s.to_string()).collect();
                return Err(SetupError::policy(format!(
                    "wizard cannot complete with unverified steps: {}",
                    names.join(",")
                )));
            }
        }
        let mut updated = self.clone();
        updated.state = to_state;
        updated.updated_at_unix_s = at_unix_s;
        updated.validate()?;
        Ok(updated)
    }

    /// Advance one step's status. VERIFIED requires a verification
    /// record; a verification record with a non-VERIFIED status is
    /// rejected before the transition check.
    pub fn advance_step(
        &self,
        step: WizardStep,
        to_status: WizardStepStatus,
        at_unix_s: u64,
        verification: Option<RemoteVerification>,
    ) -> SetupResult<Self> {
        if to_status != WizardStepStatus::Verified && verification.is_some() {
            return Err(SetupError::validation(format!(
                "step {} cannot carry a verification record with status {}",
                step, to_status
            )));
        }
        let record = self.step_record(step).ok_or_else(|| {
            SetupError::validation(format!("wizard has no record for step {}", step))
        })?;
        if !is_valid_step_transition(record.status, to_status) {
            return Err(SetupError::policy(format!(
                "invalid step transition {}: {} -> {}",
                step, record.status, to_status
            )));
        }
        if to_status == WizardStepStatus::Verified && verification.is_none() {
            return Err(SetupError::verification(format!(
                "step {} cannot become VERIFIED without a verification record",
                step
            )));
        }
        let mut updated = self.clone();
        for entry in &mut updated.steps {
            if entry.step == step {
                entry.status = to_status;
                entry.last_transition_at_unix_s = at_unix_s;
                entry.verification = verification.clone();
            }
        }
        updated.current_step = step;
        updated.updated_at_unix_s = at_unix_s;
        updated.validate()?;
        Ok(updated)
    }

    /// Verify a step remotely: the only way a step becomes VERIFIED.
    pub fn verify_step(
        &self,
        step: WizardStep,
        verification: RemoteVerification,
    ) -> SetupResult<Self> {
        if verification.verified_at_unix_s == 0 {
            return Err(SetupError::validation(
                "verified_at_unix_s must be positive",
            ));
        }
        if verification.verifier.is_empty() {
            return Err(SetupError::validation("verifier must not be empty"));
        }
        self.advance_step(
            step,
            WizardStepStatus::Verified,
            verification.verified_at_unix_s,
            Some(verification),
        )
    }
}
